book/src/proposals/durable-tx-nonces.md

```diff
@@ -0,0 +1,49 @@
+# Durable Transaction Nonces
+
+## Problem
+
+To prevent replay, Solana transactions contain a nonce field populated with a "recent" blockhash value. A transaction containing a blockhash that is too old (~2min as of this writing) is rejected by the network as invalid. Unfortunately certain use cases, such as custodial services, require a longer time allowance to produce a signature for the transaction. A mechanism is needed to enable these slow-signer network participants.
+
+## A Naive Solution
+
+An initial attempt at this was made by designating an special blockhash value that, if present in the `recent_blockhash` field, allowed the transaction to succeed so long as it additionally spent the full balance of the fee account.  However, this would introduce a security vulnerability by the fact that the transaction could be replayed by simply funding the fee account again.
+
+## A Contract-based Solution
+
+Here we propose a contract-based solution to the problem, whereby a client can "stash" a nonce value for future use in a transaction's `recent_blockhash` field. This approach is akin to the Compare and Swap atomic instruction, implemented by some CPU ISAs.
+
+### Contract Mechanics
+
+TODO: svgbob this into a flowchart, add prose
+
+```text
+Start
+Create Account
+  state = Uninitialized
+Initialize Instruction
+  if hash is recent
+    stored_hash = hash
+    state = Initialized
+    success
+  else
+    error
+Spend Instruction
+  if spend_hash == stored_hash
+    if next_hash is recent
+      stored_hash = next_hash
+      success
+    else
+      error
+  else
+    error
+```
+
+### Runtime Support
+
+The contract alone is not sufficient for implementing this feature, as an extant `recent_blockhash` is still required on any transaction executing the `Spend` instruction. To alleviate this, a `flags` field will be added to the transaction message and one bit reserved to signal the use of a Durable Transaction Nonce, which skips the typical age check.
+
+### Open Questions
+
+* Should an instruction be added to explicitly drive the account state back to `Uninitialized` or just leave it subject to rent?
```

Comment:
If you leave it subject to rent, what's stopping someone from creating that same contract and using it to replay the transaction?

Reply:
Nonce is different

Reply:
Even if the key is leaked and the account is recreated, the tx can only fire once.

Reply:
What does that mean?

Reply:
The hashes have to be in the `blockhash_queue` to qualify for stashing.

It does occur to me now that we should prevent `next_hash == spend_hash` on a `Spend` instruction to force a value change

Reply:
@t-nelson that is correct.  also removing the balance needs to come after spend_hash cannot be recreated, and the account needs to have enough rent to cover that.  It's probably easier to make them rent exempt.

Reply:
Agreed.  Does having a way to `Deinitialize`-like instruction seem like it would have value?  I can't convince myself that it does

Reply:
rent is specified such that the account lives for at least one epoch after creation, way past the range of recent_blockhashes

Reply:
> If you leave it subject to rent, what's stopping someone from creating that same contract and using it to replay the transaction?
rent is specified such that the account lives for at least one epoch after creation, way past the range of recent_blockhashes